File: checks/ruby/lib/section508/document.rb
# frozen_string_literal: true

require "nokogiri"

module Section508
  module Document
    WRAPPER_ID = "section508-root"

    module_function

    def from(source)
      case source
      when String
        parse_html(source)
      when Nokogiri::XML::Document
        source.at_css("body") || source.root
      when Nokogiri::XML::Node
        source
      else
        from_duck(source)
      end
    end

    def from_duck(source)
      if source.respond_to?(:native)
        native = source.native
        return native if native.is_a?(Nokogiri::XML::Node)
      end
      if source.respond_to?(:html)
        html = source.html
        return html if html.is_a?(Nokogiri::XML::Node)
        return parse_html(html.to_s)
      end
      if source.respond_to?(:body)
        body = source.body
        return body if body.is_a?(Nokogiri::XML::Node)
        return parse_html(body.to_s)
      end

      raise ArgumentError, "cannot run Section 508 checks on #{source.class}"
    end

    def parse_html(html)
      wrapped = %(<!DOCTYPE html><html><body><div id="#{WRAPPER_ID}">#{html}</div></body></html>)
      parse_document(wrapped).at_css("##{WRAPPER_ID}")
    end

    def parse_document(html)
      if defined?(Nokogiri::HTML5)
        Nokogiri::HTML5::Document.parse(html)
      else
        Nokogiri::HTML4::Document.parse(html)
      end
    end
  end
end

File: checks/python/src/section508/document.py
from bs4 import BeautifulSoup, Tag

ROOT_ID = "root"


def parse(html: str) -> Tag:
    """Parse a fragment the way the Node wrapper does (container div)."""
    soup = BeautifulSoup(
        f'<!DOCTYPE html><html><body><div id="{ROOT_ID}">{html}</div></body></html>',
        "html.parser",
    )
    root = soup.find(id=ROOT_ID)
    return root if root is not None else soup


def _looks_like_document(html: str) -> bool:
    head = html.lstrip()[:20].lower()
    return head.startswith("<!doctype") or head.startswith("<html")


def parse_html(html: str) -> Tag:
    if _looks_like_document(html):
        return BeautifulSoup(html, "html.parser")
    return parse(html)


def to_root(source) -> Tag:
    if isinstance(source, str):
        return parse_html(source)
    if isinstance(source, (BeautifulSoup, Tag)):
        return source
    content = getattr(source, "content", None)
    if callable(content):
        html = content()
        if not isinstance(html, str):
            html = str(html)
        return parse_html(html)
    page_source = getattr(source, "page_source", None)
    if isinstance(page_source, str):
        return parse_html(page_source)
    text = getattr(source, "text", None)
    if isinstance(text, str):
        return parse_html(text)
    raise TypeError(f"Cannot score accessibility of {type(source)!r}")
